Conditions 101747 Fix 526EZ VA Medical Records Save Functionality

williamphelps13 opened the pull request:
### Slack post with recording of before and after: https://dsva.slack.com/archives/C04KW0B46N5/p1739993034867679
- Note that `useVaCards` prop that updated the styling has since been removed

## Summary

### Summarize the changes that have been made to the platform:
#### src/applications/disability-benefits/all-claims/content/vaMedicalRecords.jsx
- Add optional chaining to fix runtime errors occurring in Cypress

#### src/applications/disability-benefits/all-claims/pages/vaMedicalRecords.js
- On this page the changes are just to the`vaTreatmentFacilities` use of the Platform ArrayField.
  - It was surprising that this page uses the Platform ArrayField because previous ArrayField exploration in the 526EZ used the All Claims ArrayField 
  - This includes the addDisabilities ArrayField which was recently updated from using the All Claims ArrayField to the Platform ArrayField
- The primary change is the addition of the ArrayField option `reviewMode: true`. 
- The issue that this resolves was unexpected behavior in which the Save button for the ArrayField did not have an effect - it would not save the item. "Saving" the item for an ArrayField is converting it from Edit mode to Review mode.
- The `addDisabilities` page where users add new conditions (and the  526EZ page with the most Active Users at any given time) has an ArrayField with a Save button that on click converts items from Edit to Review mode.
- Removing the ArrayField `showSave: true` option would remove the Save button and be an alternate solution to this issue. 
  - With this set of options the way to change an item from Edit to Review mode is to click "Add another facility". 
  - However, this set of options would not align with the addDisabilities ArrayField and the replacement of the "Save" button with the "Add another facility" as the way to change the item from Edit to Review does not seem to be as intuitive of UX.
- Adding `reviewMode: true` makes the Save button have the desired effect of changing an item from Edit to Review.
- In addition to this primary change there are two additional changes that use recently added ArrayField options which align with accessibility best practices.
- The `confirmRemove: true` adds a modal when the user clicks an item's "Remove" button to protect against accidental item removal.
- The `useVaCards: true` updated the item card to a Web Component and removes the gray background which has also been flagged as creating insufficient text contrast. However, due to the increase it scope necessary to fix the e2e testing this was removed.

#### src/applications/disability-benefits/all-claims/migrations/06-fix-treatedDisabilityNames.js
- Remove `vaTreatmentFacilities` fallback to array from 3 year old migration to fix a bug that was occurring because of this fallback.
- If `vaTreatmentFacilities` is an empty array `[]` on first time navigating to the page instead of being `undefined` then the editing state (see below code) is also an empty array `[]` by default rather then `[true]` (because an empty array `[]` is truthy). Thus the first item (which is created by default) is created in review mode rather then editing mode.
<img width="568" alt="Screenshot 2025-03-03 at 9 23 08 AM" src="https://github.com/user-attachments/assets/c11eea7c-6537-4355-8576-242032c38654" />

```
this.state = {
      editing: props.formData
        ? props.formData.map(
            (item, index) => !errorSchemaIsValid(props.errorSchema[index]),
          )
        : [true],
```

#### src/applications/disability-benefits/all-claims/tests/pages/vaMedicalRecords.unit.spec.jsx
- Fix now failing 'should not submit without all required info' test. Platform ArrayField with reviewMode: true must have data as undefined rather than an empty array to behave properly.

### Which team do you work for, does your team own the maintenance of this component?
- I work for the Conditions Team.
- This page is in the `all-claims` application within the "Supporting Evidence" chapter. Our team is primarily focused on the `Conditions` chapter. This was an issue reported from a stakeholder to our team and is a fix to the ArrayField

### What is the difference between `reviewMode` as falsey and `reviewMode: true`
- When `reviewMode` is falsey the last item is always in Edit mode and its not possible to click save to make the last item in "Review Mode" - this is potentially to help users make edits more freely. However, when `reviewMode` is turned on with the `reviewMode: true` then the last time can be in Review Mode.

#### In the render method:
```
const isReviewMode = uiOptions.reviewMode;
if (isReviewMode ? isEditing : isLast || isEditing) { ... }
```
- If `reviewMode: true`, an item is shown in its Review Mode unless explicitly edited.
- If reviewMode is omitted, the last item is always editable.

#### If reviewMode is falsey the last item remains in edit mode by default
```
handleEdit(index, status = true) {
  this.setState(set(['editing', index], status, this.state), () => {
    this.scrollToRow(`${this.props.idSchema.$id}_${index}`);
    this.focusOnFirstFocusableElement(index);
  });
}
```
- If `reviewMode: true`, items start in Review mode and require clicking "Edit" to enter Edit mode.
- If `reviewMode` is omitted, the last item remains in Edit mode by default.

## Related issue(s)
- [Adding the first VA medical record and the save action #101747](https://github.com/department-of-veterans-affairs/va.gov-team/issues/101747)

## Testing and Screenshots

### Describe what the old behavior was prior to the change:
- First input of the Supporting Evidence chapter "Is there any evidence you’d like us to review as part of your claim?" > selects "Yes"
<img width="539" alt="Screenshot 2025-02-19 at 11 49 10 AM" src="https://github.com/user-attachments/assets/2fc39233-0c08-4215-b407-18936cd2bb34" />

- Follow-up question "What type of evidence do you want us to review as part of your claim?" > checks "VA medical records" 
<img width="541" alt="Screenshot 2025-02-19 at 11 49 17 AM" src="https://github.com/user-attachments/assets/3cbdcce1-2d49-474f-b71e-86858f48a81c" />

- Clicks "Continue" button to navigate to the "VA medical records" page
- Observes gray background for the ArrayField Facility Item
<img width="563" alt="Screenshot 2025-02-19 at 11 50 02 AM" src="https://github.com/user-attachments/assets/0bcad984-e8de-422e-8059-2abfdc096f3a" />

- Adds details for a facility: "Name", "Condition", "First Visit", "Country", "State", and "City"
<img width="580" alt="Screenshot 2025-02-19 at 11 51 16 AM" src="https://github.com/user-attachments/assets/3d2c36eb-7395-414b-a95a-43e68ed51b02" />

- Clicks "Save" button > **_No change occurs_** which would be very confusing and concerning to users
<img width="589" alt="Screenshot 2025-02-19 at 11 51 30 AM" src="https://github.com/user-attachments/assets/c3ac80e5-0215-4a55-b41f-e66f4d6628a7" />

- If clicks "Add another facility" button > first Item changes from Edit to Review while second Item is created in Edit mode
<img width="586" alt="Screenshot 2025-02-19 at 11 51 57 AM" src="https://github.com/user-attachments/assets/c25b9c79-aef4-44cd-8bb9-8213bb12af06" />

- If clicks "Edit" button for item in Review mode > Item changes from Review to Edit mode
<img width="614" alt="Screenshot 2025-02-19 at 11 52 48 AM" src="https://github.com/user-attachments/assets/29480243-afe6-4ad0-b04b-d99127bbeb01" />

- If clicks "Remove" button then "Removes" item without confirmation
<img width="603" alt="Screenshot 2025-02-19 at 11 53 12 AM" src="https://github.com/user-attachments/assets/ece7376a-0732-4ea7-957f-2e29db5c40ba" />

- If click "Continue" button with invalid Item > Throws errors
<img width="612" alt="Screenshot 2025-02-19 at 11 53 29 AM" src="https://github.com/user-attachments/assets/e67b2b44-1fad-44e3-b51b-7a6c3842874d" />

- If click "Continue" button with valid Item > Hit back to return to page > Kept Item and last Item is in Edit mode 
<img width="525" alt="Screenshot 2025-02-19 at 11 54 09 AM" src="https://github.com/user-attachments/assets/466ebbda-e68b-4bf3-aedb-69d4cf4f39be" />
<img width="624" alt="Screenshot 2025-02-19 at 11 54 30 AM" src="https://github.com/user-attachments/assets/dcb75ff5-859c-46e3-8cc5-86fb3b3da14c" />

### Describe the steps required to verify your changes are working as expected:
- First input of the Supporting Evidence chapter "Is there any evidence you’d like us to review as part of your claim?" > select "Yes"
<img width="506" alt="Screenshot 2025-02-18 at 2 38 23 PM" src="https://github.com/user-attachments/assets/74de20d5-a5ea-474a-af49-3b6b6c31a5f1" />

- Follow-up question "What type of evidence do you want us to review as part of your claim?" > check "VA medical records"
<img width="506" alt="Screenshot 2025-02-18 at 2 38 34 PM" src="https://github.com/user-attachments/assets/9ef606ce-310c-4621-9dc7-122b8be704ac" />

- Click "Continue" button to navigate to the "VA medical records" page
- Observe gray background for the ArrayField Facility Item
<img width="603" alt="Screenshot 2025-02-28 at 9 42 54 AM" src="https://github.com/user-attachments/assets/8e2a1a92-a176-41e9-a986-1a1d50822269" />

- Add details for a facility: "Name", "Condition", "First Visit", "Country", "State", and "City"
<img width="608" alt="Screenshot 2025-02-28 at 9 43 32 AM" src="https://github.com/user-attachments/assets/4b7e2f19-7e59-4651-9bc0-cbbb5b1c4b0a" />

- Click "Save" button >  Item changes from Edit to Review
<img width="532" alt="Screenshot 2025-02-28 at 9 43 43 AM" src="https://github.com/user-attachments/assets/29422e81-c5c8-4b11-8396-77c027e13134" />

- If click "Add another facility" button before click "Save" button when facility is invalid > first Item changes from Edit to Review while second Item is created in Edit mode
- If click "Add another facility" button before click "Save" button when facility is valid > first Item changes from Edit to Review while second Item is created in Edit mode
<img width="621" alt="Screenshot 2025-02-28 at 9 43 57 AM" src="https://github.com/user-attachments/assets/64818c28-6b1b-42dc-8d60-a9b4bf813c53" />

- If click "Edit" button for item in Review mode > Item changes from Review to Edit mode and to convert back to Review mode click "Save" button again
<img width="595" alt="Screenshot 2025-02-28 at 9 44 29 AM" src="https://github.com/user-attachments/assets/6b445b74-28cd-4944-a141-103bb8841c30" />

- If clicks "Remove" button then a modal pops up asking user "Are you sure you want to remove this facility?" with "Yes, remove this facility" OR "No, cancel"
<img width="679" alt="Screenshot 2025-02-28 at 9 44 46 AM" src="https://github.com/user-attachments/assets/cbc466ee-8161-496e-8ddb-ee318e26c95d" />

- If click "Continue" button with invalid Item > Throws errors
<img width="642" alt="Screenshot 2025-02-28 at 9 45 07 AM" src="https://github.com/user-attachments/assets/50730bc2-33ff-434a-8514-f9b9ba01bf3a" />

- If click "Continue" button with valid Item > Hit back to return to page > Kept Item and last item is in Review mode
<img width="592" alt="Screenshot 2025-02-28 at 9 53 56 AM" src="https://github.com/user-attachments/assets/97db4c3c-11ff-4860-8547-c782088440aa" />

## Acceptance criteria

### Quality Assurance & Testing
- [x] I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [x] Linting warnings have been addressed
- [x] Documentation has been updated ([link to documentation](#) \*if necessary)
- [x] Screenshot of the developed feature is added
- [x] [Accessibility testing](https://depo-platform-documentation.scrollhelp.site/collaboration-cycle/prepare-for-an-accessibility-staging-review) has been performed

### Error Handling
- [x] Browser console contains no warnings or errors.
- [x] Events are being sent to the appropriate logging solution
- [x] Feature/bug has a monitor built into Datadog or Grafana (if applicable)

### Authentication
- [x] Did you login to a local build and verify all authenticated routes work as expected with a test user

## Release Plan (In ET)
- Merge Wednesday March 12, 2025 (to accommodate team schedules for bug bash) after daily prod deploy - UPDATE production deploy is occurring 2pm ET currently because of an issue in the deploy script in which it is not honoring daylight savings time
- Team bug bash that afternoon (not because this is a high risk deploy, but because of the lack of experience of the team in this chapter and to gain more experience in this chapter)
- Monitor Realtime Pages:
  - `/disability/file-disability-claim-form-21-526ez/supporting-evidence/va-medical-records` - modified page
    - Normal 3pm active users: 34
      - <img width="861" alt="Screenshot 2025-03-12 at 12 03 38 PM" src="https://github.com/user-attachments/assets/56882c21-c030-4948-b341-33a33c25d191" />
    - March 13, 2025 1pm active users (before release takes effect): 22
    - March 13, 2025 2pm active users: 22
    - March 13, 2025 3pm active users: 33
  - `/disability/file-disability-claim-form-21-526ez/supporting-evidence/private-medical-records` - next page if "Private medical records" is also checked
    - Normal 3pm active users: 5
      - <img width="826" alt="Screenshot 2025-03-12 at 12 03 14 PM" src="https://github.com/user-attachments/assets/9fa41abf-b798-497e-b743-2c14171e20a1" />
    - March 13, 2025 1pm active users (before release takes effect): 1
    - March 13, 2025 2pm active users: 3
    - March 13, 2025 3pm active users: 14
  - `/disability/file-disability-claim-form-21-526ez/supporting-evidence/additional-evidence` - next page if just "Supporting (lay) statements or other evidence" is also checked
    - Normal 3pm active users: 15
      - <img width="857" alt="Screenshot 2025-03-12 at 12 04 26 PM" src="https://github.com/user-attachments/assets/3d021dcc-5d1e-42a2-b66e-296dde87ecb6" />
    - March 13, 2025 1pm active users (before release takes effect): 22
    - March 13, 2025 2pm active users: 23
    - March 13, 2025 3pm active users: 23
  - `/disability/file-disability-claim-form-21-526ez/supporting-evidence/summary`- next page no other checkbox is checked
    - Normal 3pm active users: 1
      - <img width="859" alt="Screenshot 2025-03-12 at 12 05 06 PM" src="https://github.com/user-attachments/assets/4d87b824-f061-4299-b685-9d5cf11f0f0a" />
    - March 13, 2025 1pm active users (before release takes effect): 1
    - March 13, 2025 2pm active users: 3
    - March 13, 2025 3pm active users: 12
  - `/disability/file-disability-claim-form-21-526ez/how-claim-exams-work`
    - Normal 3pm active users: 3
    - March 13, 2025 1pm active users (before release takes effect): 3
    - March 13, 2025 2pm active users: 3
    - March 13, 2025 3pm active users: 13
  - `/disability/file-disability-claim-form-21-526ez/payment-information`
    - Normal 3pm active users: 10
      - <img width="860" alt="Screenshot 2025-03-12 at 12 18 25 PM" src="https://github.com/user-attachments/assets/becd1a51-40cc-4ba1-960c-30fdf435c1ce" />
    - March 13, 2025 1pm active users (before release takes effect): 11
    - March 13, 2025 2pm active users: 8
    - March 13, 2025 3pm active users: 18

williamphelps13 committed:
Fix save issue by adding reviewMode: true ArrayField option

williamphelps13 committed:
Add confirmRemove and useVaCards ArrayField options to align with addDisabilities

bacitracin reviewed (dismissed):
As always thank you for the thorough documentation of the issue!

williamphelps13 committed:
Merge branch 'main' into conditions-101747-fix-medical-records-save

williamphelps13 committed:
Add optional chaining to vaTreatmentFacilities viewField to fix runtime errors in Cypress

williamphelps13 committed:
Remove vaTreatmentFacilities fallback to array from 3 year old migration to fix tests

williamphelps13 committed:
Fix now failing 'should not submit without all required info' test

Platform ArrayField with reviewMode: true must have data as undefined rather than an empty array to behave properly.

williamphelps13 committed:
Remove useVaCards since is resulting in continually failing E2E test

williamphelps13 committed:
Merge branch 'main' into conditions-101747-fix-medical-records-save

kylesoskin commented on src/applications/disability-benefits/all-claims/migrations/06-fix-treatedDisabilityNames.js:
```diff
@@ -4,7 +4,7 @@ import { sippableId } from '../utils';
 export default function fixTreatedDisabilityNamesKey(savedData) {
   const formData = clone(savedData.formData);
-  const facilities = formData.vaTreatmentFacilities || [];
+  const facilities = formData.vaTreatmentFacilities;
```
ah I see. I didnt read enough 😄

williamphelps13 committed:
Merge branch 'main' into conditions-101747-fix-medical-records-save

williamphelps13 committed:
Merge branch 'main' into conditions-101747-fix-medical-records-save

christinec-fftc reviewed (approved):
Thanks for the video, that was super helpful! In case you want to add a unit test for the migration change, that would go in `src/applications/disability-benefits/all-claims/tests/migrations/migrations.unit.spec.js`
